File: src/schema/resumeSchema.js
export const SCHEMA_DEF = {
  identity: {
    type: 'object',
    label: 'Identity',
    fields: [
      { id: 'full_name', label: 'Full Name', required: true },
      { id: 'headline', label: 'Headline', required: false },
      { id: 'summary', label: 'Summary', required: false },
    ]
  },
  contact: {
    type: 'object',
    label: 'Contact Information',
    fields: [
      { id: 'email', label: 'Email', required: true },
      { id: 'phone', label: 'Phone', required: false },
      { id: 'location', label: 'Location', required: false },
      { id: 'linkedin', label: 'LinkedIn', required: false },
      { id: 'github', label: 'GitHub', required: false },
      { id: 'portfolio', label: 'Portfolio', required: false },
    ]
  },
  work_experience: {
    type: 'array',
    label: 'Work Experience',
    fields: [
      { id: 'company', label: 'Company', required: true },
      { id: 'title', label: 'Title', required: true },
      { id: 'location', label: 'Location', required: false },
      { id: 'start_date', label: 'Start Date', required: false },
      { id: 'end_date', label: 'End Date', required: false },
      { id: 'is_current', label: 'Is Current', required: false },
      { id: 'responsibilities', label: 'Responsibilities (Bullets)', required: false, isList: true }
    ]
  },
  education: {
    type: 'array',
    label: 'Education',
    fields: [
      { id: 'institution', label: 'Institution', required: true },
      { id: 'degree', label: 'Degree', required: false },
      { id: 'field_of_study', label: 'Field of Study', required: false },
      { id: 'graduation_year', label: 'Graduation Year', required: false },
      { id: 'gpa', label: 'GPA', required: false },
      { id: 'honors', label: 'Honors', required: false }
    ]
  },
  skills: {
    type: 'object',
    label: 'Skills',
    fields: [
      { id: 'technical', label: 'Technical Skills', required: false, isList: true },
      { id: 'soft', label: 'Soft Skills', required: false, isList: true },
      { id: 'languages', label: 'Languages', required: false, isList: true },
      { id: 'other', label: 'Other Skills', required: false, isList: true }
    ]
  },
  projects: {
    type: 'array',
    label: 'Projects',
    fields: [
      { id: 'name', label: 'Name', required: true },
      { id: 'description', label: 'Description', required: false },
      { id: 'technologies', label: 'Technologies', required: false, isList: true },
      { id: 'url', label: 'URL', required: false },
      { id: 'date', label: 'Date', required: false }
    ]
  },
  certifications: {
    type: 'array',
    label: 'Certifications',
    fields: [
      { id: 'name', label: 'Name', required: true },
      { id: 'issuer', label: 'Issuer', required: false },
      { id: 'date', label: 'Date', required: false },
      { id: 'credential_id', label: 'Credential ID', required: false },
      { id: 'url', label: 'URL', required: false }
    ]
  },
  awards: {
    type: 'array',
    label: 'Awards & Achievements',
    fields: [
      { id: 'title', label: 'Title', required: true },
      { id: 'issuer', label: 'Issuer', required: false },
      { id: 'date', label: 'Date', required: false },
      { id: 'description', label: 'Description', required: false }
    ]
  },
  publications: {
    type: 'array',
    label: 'Publications',
    fields: [
      { id: 'title', label: 'Title', required: true },
      { id: 'publisher', label: 'Publisher', required: false },
      { id: 'date', label: 'Date', required: false },
      { id: 'url', label: 'URL', required: false }
    ]
  },
  volunteer: {
    type: 'array',
    label: 'Volunteer Experience',
    fields: [
      { id: 'organisation', label: 'Organization', required: true },
      { id: 'role', label: 'Role', required: false },
      { id: 'start_date', label: 'Start Date', required: false },
      { id: 'end_date', label: 'End Date', required: false },
      { id: 'description', label: 'Description', required: false }
    ]
  }
};

/**
 * Returns a dictionary of template instances grouped by parent,
 * along with any validation warnings.
 *
 * Example Return:
 * {
 *   counts: { work_experience: 2, education: 1 },
 *   warnings: [{ type: 'missing_required', parent: 'work_experience', instanceId: 1, missing: ['title'] }],
 *   instances: {
 *     work_experience: {
 *       1: { boundFields: ['company'], objects: [ fabricObj ] },
 *       2: { boundFields: ['company', 'title'], objects: [...] }
 *     },
 *     identity: {
 *       0: { boundFields: ['full_name'], objects: [...] }
 *     }
 *   }
 * }
 */
export function analyzeCanvasBindings(canvas) {
  if (!canvas) return null;

  const instances = {};
  const counts = {};
  const warnings = [];

  const objects = canvas.getObjects();

  // 1. Group objects by parent and instance ID
  objects.forEach(obj => {
    const { bindParent, bindInstanceId = 0, bindField } = obj.customMeta || {};
    if (!bindParent || !bindField) return;

    if (!instances[bindParent]) {
      instances[bindParent] = {};
    }
    if (!instances[bindParent][bindInstanceId]) {
      instances[bindParent][bindInstanceId] = { boundFields: [], objects: [] };
    }

    if (!instances[bindParent][bindInstanceId].boundFields.includes(bindField)) {
      instances[bindParent][bindInstanceId].boundFields.push(bindField);
    }
    instances[bindParent][bindInstanceId].objects.push(obj);
  });

  // 2. Compute max counts for arrays and check required fields
  Object.keys(instances).forEach(parentKey => {
    const schemaDef = SCHEMA_DEF[parentKey];
    if (!schemaDef) return;

    if (schemaDef.type === 'array') {
      const instanceIds = Object.keys(instances[parentKey]).map(Number);
      counts[parentKey] = instanceIds.length > 0 ? Math.max(...instanceIds) : 0;
    }

    const requiredFields = schemaDef.fields.filter(f => f.required).map(f => f.id);

    // Check each instance
    Object.keys(instances[parentKey]).forEach(instId => {
      const boundFields = instances[parentKey][instId].boundFields;
      const missing = requiredFields.filter(f => !boundFields.includes(f));
      if (missing.length > 0) {
        warnings.push({
          parent: parentKey,
          instanceId: Number(instId),
          missing,
          label: schemaDef.label,
        });
      }
    });
  });

  return { instances, counts, warnings };
}

File: backend/resume_schema.py
"""
resume_schema.py
Authoritative JSON Schema for a structured resume.
Used by:
  - pipeline.py  → passed as guided_json to Llama NIM
  - validation   → jsonschema.validate() post-LLM call
"""

RESUME_SCHEMA = {
    "type": "object",
    "description": "Structured representation of a professional resume.",
    "required": ["full_name", "contact", "work_experience", "education"],
    "additionalProperties": False,
    "properties": {
        "document_name": {
            "type": ["string", "null"],
            "description": "User-provided name for this resume document."
        },
        "original_upload_filename": {
            "type": ["string", "null"],
            "description": "Original filename if uploaded, or 'nan'."
        },
        "profile_pic": {
            "type": ["string", "null"],
            "description": "URL of the uploaded profile picture. Optional."
        },
        "template_id": {
            "type": ["integer", "null"],
            "description": "ID of the selected resume template. Optional."
        },

        # ── IDENTITY ──────────────────────────────────────────────
        "full_name": {
            "type": "string",
            "description": "Candidate's full legal or professional name. REQUIRED."
        },
        "headline": {
            "type": ["string", "null"],
            "description": "Short professional tagline or title. Optional."
        },
        "summary": {
            "type": ["string", "null"],
            "description": "Career summary or objective paragraph. Optional."
        },

        # ── CONTACT ───────────────────────────────────────────────
        "contact": {
            "type": "object",
            "required": ["email"],
            "additionalProperties": False,
            "properties": {
                "email":       {"type": "string"},
                "phone":       {"type": ["string", "null"]},
                "location":    {"type": ["string", "null"]},
                "linkedin":    {"type": ["string", "null"]},
                "github":      {"type": ["string", "null"]},
                "portfolio":   {"type": ["string", "null"]},
                "other_links": {"type": "array", "items": {"type": "string"}}
            }
        },

        # ── WORK EXPERIENCE ───────────────────────────────────────
        "work_experience": {
            "type": "array",
            "items": {
                "type": "object",
                "required": ["company", "title"],
                "additionalProperties": False,
                "properties": {
                    "company":          {"type": "string"},
                    "title":            {"type": "string"},
                    "location":         {"type": ["string", "null"]},
                    "start_date":       {"type": ["string", "null"]},
                    "end_date":         {"type": ["string", "null"]},
                    "is_current":       {"type": ["boolean", "null"]},
                    "responsibilities": {"type": "array", "items": {"type": "string"}}
                }
            }
        },

        # ── EDUCATION ─────────────────────────────────────────────
        "education": {
            "type": "array",
            "items": {
                "type": "object",
                "required": ["institution"],
                "additionalProperties": False,
                "properties": {
                    "institution":    {"type": "string"},
                    "degree":         {"type": ["string", "null"]},
                    "field_of_study": {"type": ["string", "null"]},
                    "graduation_year":{"type": ["string", "null"]},
                    "gpa":            {"type": ["string", "null"]},
                    "honors":         {"type": ["string", "null"]}
                }
            }
        },

        # ── SKILLS ────────────────────────────────────────────────
        "skills": {
            "type": "object",
            "additionalProperties": False,
            "properties": {
                "technical": {"type": "array", "items": {"type": "string"}},
                "soft":      {"type": "array", "items": {"type": "string"}},
                "languages": {"type": "array", "items": {"type": "string"}},
                "other":     {"type": "array", "items": {"type": "string"}}
            }
        },

        # ── PROJECTS ──────────────────────────────────────────────
        "projects": {
            "type": "array",
            "items": {
                "type": "object",
                "required": ["name"],
                "additionalProperties": False,
                "properties": {
                    "name":         {"type": "string"},
                    "description":  {"type": ["string", "null"]},
                    "technologies": {"type": "array", "items": {"type": "string"}},
                    "url":          {"type": ["string", "null"]},
                    "date":         {"type": ["string", "null"]}
                }
            }
        },

        # ── CERTIFICATIONS ────────────────────────────────────────
        "certifications": {
            "type": "array",
            "items": {
                "type": "object",
                "required": ["name"],
                "additionalProperties": False,
                "properties": {
                    "name":          {"type": "string"},
                    "issuer":        {"type": ["string", "null"]},
                    "date":          {"type": ["string", "null"]},
                    "credential_id": {"type": ["string", "null"]},
                    "url":           {"type": ["string", "null"]}
                }
            }
        },

        # ── AWARDS & ACHIEVEMENTS ─────────────────────────────────
        "awards": {
            "type": "array",
            "items": {
                "type": "object",
                "required": ["title"],
                "additionalProperties": False,
                "properties": {
                    "title":       {"type": "string"},
                    "issuer":      {"type": ["string", "null"]},
                    "date":        {"type": ["string", "null"]},
                    "description": {"type": ["string", "null"]}
                }
            }
        },

        # ── PUBLICATIONS ──────────────────────────────────────────
        "publications": {
            "type": "array",
            "items": {
                "type": "object",
                "required": ["title"],
                "additionalProperties": False,
                "properties": {
                    "title":     {"type": "string"},
                    "publisher": {"type": ["string", "null"]},
                    "date":      {"type": ["string", "null"]},
                    "url":       {"type": ["string", "null"]}
                }
            }
        },

        # ── VOLUNTEER ─────────────────────────────────────────────
        "volunteer": {
            "type": "array",
            "items": {
                "type": "object",
                "required": ["organisation"],
                "additionalProperties": False,
                "properties": {
                    "organisation": {"type": "string"},
                    "role":         {"type": ["string", "null"]},
                    "start_date":   {"type": ["string", "null"]},
                    "end_date":     {"type": ["string", "null"]},
                    "description":  {"type": ["string", "null"]}
                }
            }
        },

        # ── META ──────────────────────────────────────────────────
        "meta": {
            "type": "object",
            "additionalProperties": False,
            "properties": {
                "parse_confidence":  {"type": ["number", "null"]},
                "detected_language": {"type": ["string", "null"]},
                "total_pages":       {"type": ["integer", "null"]}
            }
        }
    }
}
